src/ops/constant/typecheck.rs

```diff
@@ -136,25 +131,16 @@ pub fn typecheck_const(typ: &ClassicType, val: &ConstValue) -> Result<(), ConstT
                 }
                 Ok(())
             }
-            (Container::Tuple(_), _) => {
-                Err(ConstTypeError::TypeMismatch(ty.clone(), tm.const_type()))
-            }
-            (Container::Sum(row), ConstValue::Sum { tag, variants, val }) => {
+            (Container::Tuple(_), _) => Err(ConstTypeError::ValueCheckFail(ty.clone(), tm.clone())),
+            (Container::Sum(row), ConstValue::Sum { tag, val }) => {
                 if tag > &row.len() {
                     return Err(ConstTypeError::InvalidSumTag);
                 }
-                if **row != *variants {
-                    return Err(ConstTypeError::TypeRowMismatch(
-                        *row.clone(),
-                        variants.clone(),
-                    ));
-                }
-                let ty = variants.get(*tag).unwrap();
+
+                let ty = row.get(*tag).unwrap();
```

Comment:
nit - slight preference to `match row.get(*tag)` to combine the check+unwrap into one